tests/UnusedCodeTest.php

```diff
@@ -779,6 +779,120 @@ function reverse(string $string): string
                         return strrev($string);
                     }'
             ],
+            'unusedByReferenceFunctionCall' => [
+                '<?php
+                    /**
+                     * @pure
+                     */
+                    function bar(string &$str): string
```

Comment:
Uhh this is definitely not pure though. It may be allowed, but it shouldn't (and I don't like to add tests that future code changes will force me to fix)

Reply:
Personally i don't see the problem with it being pure, as i don't think by-ref arguments are a side effects.

but, if you want to consider functions with by-ref arguments as non pure, i think i can close this PR and work on another to forbid byref in pure context.

Reply:
I _sort_ of see your point insofar as byref arguments could be viewed as outputs.

_However_ I think that the return type should always be the primary location for output in pure functions. I don't want to encourage

```php
function some_pure_function(string &$s) : void {}
```
over
```php
function some_pure_function(string $s) : string {}
```

– but I'm fine with 

```php
function some_pure_function(string &$s) : string {}
```

Basically I think a pure function should either _always_ exit, or _always_ return a (non-`void`) value.

LMK if you think that's too-harsh a constraint.

Reply:
i agree with you, personally i don't have any functions that behave either way, but we need to have correct errors in place.

for
```
function some_pure_function(string &$s) : void {}
```

i think it's best to show something like `A pure function must not have a 'void' return type.` ( not sure about the correct issue for this ) 

as for the other, they seem okay to me and there shouldn't be any type of error result from dismissing their result ( in case you only need to the by-reference argument result ) in which this PR fixes.

so i guess we can merge this + i will create another PR to forbid having pure functions with `void` return type.

Reply:
Sorry for commenting on a decision which has already been work done upon.

But if this isn't allowed: https://psalm.dev/r/cca607305c

I really don't get why this should be allowed: https://psalm.dev/r/a9e67c6ce9

I'd argue that functions that return something *and* have mutated passed-by-reference arguments are more confusing than functions that only do either of those.